lib/connection_controller.go

```diff
@@ -330,6 +351,146 @@ func (cc *ConnectionController) connectValidators(activeValidatorsMap *collectio
+		} else {
+			attemptedOutboundRemoteNodes = append(attemptedOutboundRemoteNodes, rn)
+		}
+	}
+
+	// Having separated the attempted and connected remote nodes, we can now find the actual number of attempted and
+	// connected remote nodes. We can then find out how many remote nodes we need to disconnect.
+	numOutboundRemoteNodes = uint32(len(attemptedOutboundRemoteNodes) + len(connectedOutboundRemoteNodes))
+	excessiveOutboundRemoteNodes := uint32(0)
+	if numOutboundRemoteNodes > cc.targetNonValidatorOutboundRemoteNodes {
+		excessiveOutboundRemoteNodes = numOutboundRemoteNodes - cc.targetNonValidatorOutboundRemoteNodes
+	}
+
+	// First disconnect the attempted remote nodes.
+	for _, rn := range attemptedOutboundRemoteNodes {
+		if excessiveOutboundRemoteNodes == 0 {
+			break
+		}
+		cc.rnManager.Disconnect(rn)
+		excessiveOutboundRemoteNodes--
+	}
+	// Now disconnect the connected remote nodes, if we still have too many remote nodes.
+	for _, rn := range connectedOutboundRemoteNodes {
+		if excessiveOutboundRemoteNodes == 0 {
+			break
+		}
+		cc.rnManager.Disconnect(rn)
+		excessiveOutboundRemoteNodes--
+	}
+}
+
+// refreshNonValidatorInboundIndex is called periodically by the non-validator connector. It is responsible for
+// disconnecting excess inbound remote nodes.
+func (cc *ConnectionController) refreshNonValidatorInboundIndex() {
+	// First let's check if we have an excess number of inbound remote nodes. If we do, we'll disconnect some of them.
+	numConnectedInboundRemoteNodes := uint32(cc.rnManager.GetNonValidatorInboundIndex().Count())
+	excessiveInboundRemoteNodes := uint32(0)
+	if numConnectedInboundRemoteNodes > cc.targetNonValidatorInboundRemoteNodes {
+		excessiveInboundRemoteNodes = numConnectedInboundRemoteNodes - cc.targetNonValidatorInboundRemoteNodes
+	}
+	// Disconnect random inbound non-validators if we have too many of them.
+	inboundRemoteNodes := cc.rnManager.GetNonValidatorInboundIndex().GetAll()
+	for _, rn := range inboundRemoteNodes {
+		if excessiveInboundRemoteNodes == 0 {
+			break
+		}
+		cc.rnManager.Disconnect(rn)
+		excessiveInboundRemoteNodes--
+	}
+}
+
+func (cc *ConnectionController) connectNonValidators() {
+	// Only connect to addresses from the addrmgr if we don't specify --connect-ips. These addresses are *not* persistent,
+	// meaning if we disconnect from one we'll try a different one.
+	// TODO: We used this condition in the old code to prevent the node from connecting to non-connect-ips nodes.
+	// 	I'm not sure whether this is still necessary. I suppose the concern here was that the connect-ips nodes
+	//	should be prioritized over the addrmgr nodes, especially during syncing. However, I think we can achieve the
+	// 	same result by defining another flag, like a boolean --sync-from-persistent-peers-only, which could indicate that
+	//	we disregard non-persistent non-connect-ips nodes during syncing, if the flag is set to true.
+	// FIXME: This about uncommenting the below condition.
```

Comment:
@diamondhands0 will have to make the call on this. I think there are two flags: `--add-ips` and `--connect-ips`. My understanding is that `--connect-ips` means ONLY connect to these peers and no other peers. I don't know if that'll work for PoS.

Reply:
@AeonSw4n - if I understand correctly, you want to be able to dial outbound connections to ips defined in the addrmgr even when the `--connect-ips` flag is defined. And you want to define the separate `--sync-from-persistent-peers-only` to prevent the node from syncing from the addrmgr peers.

What is the purpose of dialing an outbound non-validator connection from the addrmgr if you don't intend to sync from the peer?

Reply:
We discussed this off-GH with @tholonious and figured out that the `--sync-from-persistent-peers-only` flag is redundant. The `--connect-ips` flag is sufficient to maintain the syncing behavior of the old code.